lib/turbostreamer/encoders/oj.rb

```diff
@@ -0,0 +1,131 @@
+require "oj"
+
+class TurboStreamer
+  class OjEncoder
+
+    def initialize(io, options={})
+      @stack = []
+      @indexes = []
+
+      @options = options
+
+      @output = io
+      @stream_writer = ::Oj::StreamWriter.new(
+        io,
+        {mode: :json}.merge(options),
+      )
+    end
+
+    attr_reader :options
+    attr_reader :output
+    attr_reader :stream_writer
+
+    def key(k)
+      stream_writer.push_key(k)
+    end
+
+    def value(v)
+      if @stack.last == :array || @stack.last == :map
+        @indexes[-1] += 1
+      end
+      stream_writer.push_value(v)
+    end
+
+    def map_open
+      @stack << :map
+      @indexes << 0
+      stream_writer.push_object
+    end
+
+    def map_close
+      @indexes.pop
+      @stack.pop
+      stream_writer.pop
+    end
+
+    def array_open
+      @stack << :array
+      @indexes << 0
+      stream_writer.push_array
+    end
+
+    def array_close
+      @indexes.pop
+      @stack.pop
+      stream_writer.pop
+    end
+
+    def inject(string)
+      stream_writer.flush
+
+      if @stack.last == :array
+        # self.output.write(',') if @indexes.last > 0
+        @indexes[-1] += 1
+      elsif @stack.last == :map
+        # self.output.write(',') if @indexes.last > 0
+        @indexes[-1] += 1
+      end
+
+      # For string containing key and value like `"key":"value"`
+      # OJ stream writer does NOT allow writing the whole string
+      # But instead requiring key and value to be input separately
+      key_value_match_data = /\A"(\w+?)":(.+)\Z/.match(string)
+      if key_value_match_data
+        stream_writer.push_json(
+          key_value_match_data[2],
+          key_value_match_data[1],
+        )
+      else
+        stream_writer.push_json(string)
+      end
+    end
+
+    def capture(to=nil)
+      stream_writer.flush
+
+      old_writer = self.stream_writer
+      old_output = self.output
+      @indexes << 0
+
+      @output = (to || ::StringIO.new)
+      @stream_writer = ::Oj::StreamWriter.new(
```

Comment:
I think it makes no difference since the code only use last one `stack` & `indexes`
I am not sure how fast/slow `pop_all` is, but let's decide that later (until we some benchmark that proves one is clearly faster than another one)

For now we should make the test pass so it's working first
Then compare different encoder's performance with jbuilder